File: src/resources/pstnPython/sipmediaapp.py
import boto3
import logging
import os

logger = logging.getLogger()
try:
    log_level = os.environ["LogLevel"]
    if log_level not in ["INFO", "DEBUG"]:
        log_level = "INFO"
except BaseException:
    log_level = "INFO"
logger.setLevel(log_level)

chime = boto3.client("chime")
ssm = boto3.client("ssm")


def create_sip_media_app(uid, region=None, name=None, endpoint=None, **kwargs):

    logger.info(f"Creating SIP media application: {uid}")
    try:
        sip_media_app_id = chime.create_sip_media_application(
            awsRegion=region, Name=name, Endpoints=[{"LambdaArn": endpoint}]
        )["SipMediaApplication"]["SipMediaApplicationId"]
    except Exception as e:
        error = {"error": f"Exception thrown: {e}"}
        logger.error(error)
        raise RuntimeError(error)

    try:
        ssm.put_parameter(
            Name="/chime/sipMediaApp/" + uid,
            Description="sipMediaAppId",
            Overwrite=True,
            Value=sip_media_app_id,
            Type="String",
        )
    except Exception as e:
        error = {"error": f"Exception thrown: {e}"}
        logger.error(error)
        raise RuntimeError(error)

    return sip_media_app_id


def delete_sip_media_app(uid):
    logger.info(f"Deleting SIP media app: {uid}")
    try:
        sip_media_app_to_delete = ssm.get_parameter(Name="/chime/sipMediaApp/" + str(uid),)[
            "Parameter"
        ]["Value"]
    except Exception as e:
        error = {"error": f"Exception thrown: {e}"}
        logger.error(error)
        raise RuntimeError(error)

    try:
        logger.info(f"Deleting Parameter: {uid}")
        ssm.delete_parameter(
            Name="/chime/sipMediaApp/" + str(uid),
        )
    except Exception as e:
        error = {"error": f"Exception thrown: {e}"}
        logger.error(error)
        raise RuntimeError(error)

    try:
        logger.info(f"Deleting SIP media application: {uid}")
        chime.delete_sip_media_application(SipMediaApplicationId=sip_media_app_to_delete)
    except Exception as e:
        error = {"error": f"Exception thrown: {e}"}
        logger.error(error)
        raise RuntimeError(error)

    return True

File: src/resources/pstn/sipMediaApp.ts
/* eslint-disable import/no-extraneous-dependencies */
import {
  ChimeSDKVoiceClient,
  CreateSipMediaApplicationCommand,
  CreateSipMediaApplicationCommandInput,
  CreateSipMediaApplicationCommandOutput,
  DeleteSipMediaApplicationCommand,
  PutSipMediaApplicationAlexaSkillConfigurationCommand,
  PutSipMediaApplicationAlexaSkillConfigurationCommandInput,
  PutSipMediaApplicationAlexaSkillConfigurationCommandOutput,
  PutSipMediaApplicationLoggingConfigurationCommand,
  PutSipMediaApplicationLoggingConfigurationCommandInput,
  PutSipMediaApplicationLoggingConfigurationCommandOutput,
} from '@aws-sdk/client-chime-sdk-voice';

import {
  SSMClient,
  DeleteParameterCommand,
  GetParameterCommand,
  GetParameterCommandOutput,
  PutParameterCommand,
} from '@aws-sdk/client-ssm';

const chimeSDKVoiceClient = new ChimeSDKVoiceClient({
  region: process.env.AWS_REGION,
});

const ssmClient = new SSMClient({ region: process.env.AWS_REGION });

let createSipMediaApplicationResponse: CreateSipMediaApplicationCommandOutput;
let createSipMediaApplicationParams: CreateSipMediaApplicationCommandInput;
let getParameterCommandOutput: GetParameterCommandOutput;
let putSipMediaApplicationAlexaSkillInput: PutSipMediaApplicationAlexaSkillConfigurationCommandInput;
let putSipMediaApplicationLoggingInput: PutSipMediaApplicationLoggingConfigurationCommandInput;
let putSipMediaApplicationAlexaSkillOutput: PutSipMediaApplicationAlexaSkillConfigurationCommandOutput;
let putSipMediaApplicationLoggingOutput: PutSipMediaApplicationLoggingConfigurationCommandOutput;

export interface CreateSIPMediaApplicationProps {
  name?: string;
  region?: string;
  endpoint?: string;
}

export const CreateSipMediaApplication = async (
  uid: string,
  props: CreateSIPMediaApplicationProps,
) => {
  console.log(`Creating SIP media application rule: ${uid}`);
  console.log(`Create SIP media application props: ${JSON.stringify(props)}`);

  createSipMediaApplicationParams = {
    Name: props.name,
    AwsRegion: props.region,
    Endpoints: [{ LambdaArn: props.endpoint }],
  };

  console.log(
    `Create SIP media application params: ${JSON.stringify(
      createSipMediaApplicationParams,
    )}`,
  );
  try {
    createSipMediaApplicationResponse = await chimeSDKVoiceClient.send(
      new CreateSipMediaApplicationCommand(createSipMediaApplicationParams),
    );
    console.log(
      `createSipMediaApplicationResponse: ${JSON.stringify(
        createSipMediaApplicationResponse,
      )}`,
    );
  } catch (error) {
    if (error instanceof Error) {
      console.error(error);
      throw error;
    }
  }

  try {
    await ssmClient.send(
      new PutParameterCommand({
        Name: '/chime/sipMediaApp' + uid,
        Value:
          createSipMediaApplicationResponse.SipMediaApplication
            ?.SipMediaApplicationId,
        Description: 'SIP Media Application ID',
        Overwrite: true,
        Type: 'String',
      }),
    );
  } catch (error) {
    if (error instanceof Error) {
      console.error(error);
      throw error;
    }
  }

  return {
    sipMediaAppId:
      createSipMediaApplicationResponse.SipMediaApplication
        ?.SipMediaApplicationId,
  };
};

export const DeleteSipMediaApplication = async (uid: string) => {
  try {
    getParameterCommandOutput = await ssmClient.send(
      new GetParameterCommand({ Name: '/chime/sipMediaApp' + uid }),
    );
    if (
      getParameterCommandOutput.Parameter &&
      getParameterCommandOutput.Parameter.Value
    ) {
    }
  } catch (error) {
    if (error instanceof Error) {
      console.error(error);
      throw error;
    }
  }

  try {
    await chimeSDKVoiceClient.send(
      new DeleteSipMediaApplicationCommand({
        SipMediaApplicationId: getParameterCommandOutput.Parameter?.Value,
      }),
    );
    await ssmClient.send(
      new DeleteParameterCommand({ Name: '/chime/sipMediaApp' + uid }),
    );
  } catch (error) {
    if (error instanceof Error) {
      console.error(error);
      throw error;
    }
  }
};

export interface PutSipMediaApplicationLoggingProps {
  sipMediaAppId?: string;
  sipMediaApplicationLoggingConfiguration?: {
    sipMediaApplicationLoggingConfiguration: {
      enableSipMediaApplicationMessageLogs: boolean;
    };
  };
}

export const PutSipMediaApplicationLogging = async (
  props: PutSipMediaApplicationLoggingProps,
) => {
  console.log(
    `Put SIP media application logging configuration: ${JSON.stringify(props)}`,
  );

  putSipMediaApplicationLoggingInput = {
    SipMediaApplicationId: props.sipMediaAppId,
    SipMediaApplicationLoggingConfiguration: {
      EnableSipMediaApplicationMessageLogs:
        props.sipMediaApplicationLoggingConfiguration
          ?.sipMediaApplicationLoggingConfiguration
          ?.enableSipMediaApplicationMessageLogs,
    },
  };
  try {
    putSipMediaApplicationLoggingOutput = await chimeSDKVoiceClient.send(
      new PutSipMediaApplicationLoggingConfigurationCommand(
        putSipMediaApplicationLoggingInput,
      ),
    );
  } catch (error) {
    if (error instanceof Error) {
      console.error(error);
      throw error;
    }
  }

  return {
    putSipMediaApplicationLoggingConfiguration:
      putSipMediaApplicationLoggingOutput.SipMediaApplicationLoggingConfiguration,
  };
};

enum AlexaSkillStatus {
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
}

export interface PutSipMediaApplicationAlexaSkillProps {
  sipMediaApplicationAlexaSkillConfiguration?: {
    alexaSkillIds: string[];
    alexaSkillStatus: AlexaSkillStatus;
  };
  sipMediaAppId?: string;
}

export const PutSipMediaApplicationAlexaSkill = async (
  props: PutSipMediaApplicationAlexaSkillProps,
) => {
  console.log(
    `Put SIP media application Alexa skill configuration: ${JSON.stringify(
      props,
    )}`,
  );

  putSipMediaApplicationAlexaSkillInput = {
    SipMediaApplicationId: props.sipMediaAppId,
    SipMediaApplicationAlexaSkillConfiguration: {
      AlexaSkillIds:
        props.sipMediaApplicationAlexaSkillConfiguration?.alexaSkillIds,
      AlexaSkillStatus:
        props.sipMediaApplicationAlexaSkillConfiguration?.alexaSkillStatus,
    },
  };

  try {
    putSipMediaApplicationAlexaSkillOutput = await chimeSDKVoiceClient.send(
      new PutSipMediaApplicationAlexaSkillConfigurationCommand(
        putSipMediaApplicationAlexaSkillInput,
      ),
    );
  } catch (error) {
    if (error instanceof Error) {
      console.error(error);
      throw error;
    }
  }

  return {
    putSipMediaApplicationAlexaSkillConfiguration:
      putSipMediaApplicationAlexaSkillOutput.SipMediaApplicationAlexaSkillConfiguration,
  };
};
